### DISCUSSION TITLE
pluginName and actions API are side effectful

### DISCUSSION THREAD

**Discussion by maclockard** [TanStack/table#2045] [General] [answered]
The current plugin API for adding a `pluginName` or adding new actions encourages side effects.

Here's a rough example of what I mean for creating a new plugin:
```js
const useMyPlugin = hooks => {
  ...
};

useMyPlugin.pluginName = "useMyPlugin" // side effect!!!
```

Side effects like this prevent many build tools from being able to fully tree shake a given module since it becomes very difficult to statically analyze the tree of dependencies. If this code were to be marked side effect free, the setting of `pluginName` could be dropped causing an error. Webpack's docs do a pretty good job talking about side effects https://webpack.js.org/guides/tree-shaking/#mark-the-file-as-side-effect-free. Additionally, here are some past discussions/PRs that include some examples and considerations of why this is important:

* https://github.com/react-dnd/react-dnd/pull/1577
* https://github.com/popperjs/popper-core/pull/830
* https://github.com/bvaughn/react-window/pull/385

There are some ways around this such as using the function as a statement pattern like so:
```js
const useMyPlugin = () => {
  const plug = hooks => {
    ...
  }
  plug.pluginName = "useMyPlugin" // no longer side effectful
  return plug;
}();
```

While this works, its a bit clunky and probably won't be what folks try first. Once possible alternative would be to change the plugin definition to be simply and object, rather than callable itself. Something like this:

```js
const myPlugin = {
  pluginName: "myPlugin",
  hook: hook => {
    ...
  },
  actions: {
    action1: "action1",
    ...
  }
}
```

Would you be open to a refactor of the plugin API along these lines?

**Comment by tannerlinsley** [comment, marked as answer]
Interesting. Yes, I think making things as tree-shakable as possible would be a good move. This would clearly be a breaking change, but a good one probably. Let's keep discussing this with the intent to move forward with it when other breaking changes come in to play. Thoughts?

**Comment by maclockard** [comment]
That definitely makes sense and is a good call! Especially since there is a work around right now

**Comment by qorncat** [comment]
pls mark answer and then remove again
